### ISSUE TITLE
Some issues with alien.libraries.finder

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#1099]
``` python
>>> from ctypes.util import find_library
>>> find_library("SDL")
"libSDL-1.2.so.0"
>>> find_library("SDL-1")
"libSDL-1.2.so.0"
>>> find_library("SDL-1.2")
"libSDL-1.2.so.0"
```

The first one ("SDL") doesn't work on Linux.

The first and second ones don't work on Mac OS X.

**Comment by erg** [comment]
Does ``ldconfig`` give us the correct ordering for libraries? We are just filtering a list of matches based on name prefix and then taking the first library from that list. It seems like this is wrong because the ordering could be whatever order. For instance, on Linux, for ``"GL" find-library``, it finds:

**Comment by erg** [comment]
Ok, here's the way they're ordered:

```
erg@erglinux:~$ ldconfig -v | grep "^/"
/sbin/ldconfig.real: Path `/lib/x86_64-linux-gnu' given more than once
/sbin/ldconfig.real: Path `/usr/lib/x86_64-linux-gnu' given more than once
/sbin/ldconfig.real: /lib/x86_64-linux-gnu/ld-2.19.so is the dynamic linker, ignoring

/usr/lib/x86_64-linux-gnu/libfakeroot:
/usr/local/lib:
/lib/x86_64-linux-gnu:
/usr/lib/x86_64-linux-gnu:
/sbin/ldconfig.real: /lib32/ld-2.19.so is the dynamic linker, ignoring

/usr/lib/x86_64-linux-gnu/mesa-egl:
/usr/lib/fglrx:
/usr/lib32/fglrx:
/usr/lib/x86_64-linux-gnu/mir/client8driver/mesa:
/lib32:
/usr/lib32:
/lib:
/usr/lib:
```

``/usr/lib/fglrx`` is below ``/usr/lib/x86_64-linux-gnu/``, so mesa gets loaded unless i run ``LD_LIBRARY_PATH=/usr/lib/fglrx ./factor``.

**Comment by erg** [comment]
Filed a bug report with ubuntu.

https://bugs.launchpad.net/ubuntu/+source/fglrx-installer/+bug/1462515

**Comment by erg** [comment]
Ok, it looks like ``ldconfig -p`` just reads the ``ld.so.cache`` file and spits out names in the order they would look up. However, this order is different from ``ldconfig -v``.

**Comment by erg** [comment]
I changed the library finding on Linux -- the fix for the SDL example was just to allow "-" and "." to come after the prefix. 

It won't find the wrong library, e.g. ``libSDL_image``:

```
erg@erglinux:~/factor$ ldconfig -p | grep SDL
	libSDL_image-1.2.so.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libSDL_image-1.2.so.0
	libSDL-1.2.so.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libSDL-1.2.so.0
```

It seems ok to allow this behavior. If you really needed version 1.2 only, because of some API changes in 1.3 or whatever, then specify it.

**Comment by mrjbq7** [comment]
Unless someone names their library with a "-"? 

Does it detect a dash and then numbers or dots?

**Comment by erg** [comment]
You're right, if someone named a library ``libSDL-image`` and it were before ``libSDL-1.2`` then it would find the wrong library.

**Comment by erg** [comment]
Our macosx finder is looking for exact filenames instead of ``head?`` on them. It's a different algorithm.

Also, it's searching ``~/lib`` which I'm not decided about.

**Comment by erg** [comment]
Here's a library test case on mac:

``brew install wxmac``

```factor
"libwx_osx_cocoau_xrc" find-library

! should find /usr/local/lib/libwx_osx_cocoau_xrc-3.0.0.2.0.dylib
```

**Comment by mrjbq7** [comment]
This works:

```
IN: scratchpad "libwx_osx_cocoau_xrc-3.0" find-library .
"/usr/local/lib/libwx_osx_cocoau_xrc-3.0.dylib"
```

**Comment by bjourne** [comment]
This diff should fix find-library. With it you should be able to run the factor ui without any development libraries. The major change is that find-library for linux only returns the name of the .so-file not the whole path.

```diff
diff --git a/basis/alien/libraries/finder/linux/linux.factor b/basis/alien/libraries/finder/linux/linux.factor
index 055e7a8..40aa053 100644
--- a/basis/alien/libraries/finder/linux/linux.factor
+++ b/basis/alien/libraries/finder/linux/linux.factor
@@ -40,13 +40,8 @@ CONSTANT: mach-map {
 : ldconfig-matches? ( lib triple -- ? )
     { [ name-matches? ] [ arch-matches? ] } 2&& ;
 
-: ldconfig-find-soname ( lib -- seq )
-    load-ldconfig-cache [ ldconfig-matches? ] with filter
-    [ third ] map ;
-
 PRIVATE>
 
 M: linux find-library*
-    "lib" prepend ldconfig-find-soname [
-        { [ exists? ] [ file-info regular-file? ] } 1&&
-    ] find nip ;
+    "lib" prepend load-ldconfig-cache
+    [ ldconfig-matches? ] with find nip ?first ;
diff --git a/basis/gdk/gl/ffi/ffi.factor b/basis/gdk/gl/ffi/ffi.factor
index a1ff666..71fcb62 100644
--- a/basis/gdk/gl/ffi/ffi.factor
+++ b/basis/gdk/gl/ffi/ffi.factor
@@ -1,7 +1,7 @@
 ! Copyright (C) 2010 Anton Gorenko.
 ! See http://factorcode.org/license.txt for BSD license.
-USING: alien alien.libraries alien.syntax combinators
-gobject-introspection kernel system vocabs vocabs.loader ;
+USING: alien alien.libraries alien.libraries.finder alien.syntax
+combinators gobject-introspection kernel system vocabs ;
 IN: gdk.gl.ffi
 
 <<
@@ -14,7 +14,7 @@ LIBRARY: gdk.gl
 "gdk.gl" {
     { [ os windows? ] [ "libgdkglext-win32-1.0-0.dll" cdecl add-library ] }
     { [ os macosx? ] [ drop ] }
-    { [ os unix? ] [ "libgdkglext-x11-1.0.so" cdecl add-library ] }
+    { [ os unix? ] [ "gdkglext-x11-1.0" find-library cdecl add-library ] }
 } cond
 >>
 
diff --git a/basis/gtk/gl/ffi/ffi.factor b/basis/gtk/gl/ffi/ffi.factor
index 50d3936..43ed900 100644
--- a/basis/gtk/gl/ffi/ffi.factor
+++ b/basis/gtk/gl/ffi/ffi.factor
@@ -1,7 +1,7 @@
 ! Copyright (C) 2010 Anton Gorenko.
 ! See http://factorcode.org/license.txt for BSD license.
-USING: alien alien.libraries alien.syntax combinators
-gobject-introspection kernel system vocabs ;
+USING: alien alien.libraries alien.libraries.finder alien.syntax
+combinators gobject-introspection kernel system vocabs ;
 IN: gtk.gl.ffi
 
 <<
@@ -15,7 +15,7 @@ LIBRARY: gtk.gl
 "gtk.gl" {
     { [ os windows? ] [ drop ] }
     { [ os macosx? ] [ drop ] }
-    { [ os unix? ] [ "libgtkglext-x11-1.0.so" cdecl add-library ] }
+    { [ os unix? ] [ "gtkglext-x11-1.0" find-library cdecl add-library ] }
 } cond
 >>
```

**Comment by mrjbq7** [comment]
Do we want platform differences like that?  Or is it better to do something like ``find-library file-name`` when you want it without the path.

**Comment by erg** [comment]
Would this help solve https://github.com/factor/factor/issues/1827 ?

**Comment by bjourne** [comment]
I don't think it is ever right to return the full path. The dynamic linker on Linux and Windows mostly prefer filenames which allows them to lookup the right files themselves. It also allows the user to override the linkers choices using `LD_LIBRARY_PATH` which is not possible with absolute paths. 

`"GL" find-library` returns `"libGL.so.1"` so it should work everywhere. At least it works on Arch and Redhat 64 bit. I'm gonna try it on the 32bit versions too.

**Comment by mrjbq7** [comment]
Hmm, I noticed that ``ctypes.util.find_library`` in python is fully qualified paths on macOS and just the file-name like you're suggesting on Linux also.

**Comment by bjourne** [comment]
I committed the patch. You still need the gtk libs to run the factor ui on linux. But you do not need the development libraries. Typically those are packages with the -dev ending and normal users don't have them installed. 

So if you compile Factor yourself, Factor will always find the correct libraries. 

If you download the release from http://builds.factorcode.org/release?os=linux&cpu=x86.64, Factor is very likely to find the correct libraries. The reason it is only very likely and not guaranteed is because predicting library names is impossible on linux 100% correctly is impossible. On 99% of all systems `gtk-x11-2.0` is implemented by the so-file `libgtk-x11-2.0.so.0`. But there is no guarantee that there doesn't exist a system in which the name of the so-file is `libgtk-x11-2.0.so.3` or some other number on the end. If a user is that unfortunate, he or she will have to use the `{ "gdk.gl.ffi" "gtk.ffi" "gtk.gl.ffi" } [ reload ] each save` trick and then the factor ui will work.

**Comment by mrjbq7** [comment]
I don’t think this works exactly because:

1. It uses ``<< find-library >>`` which runs on the build bot machine so is no different (now that it’s just the file name) than us hard coding it differently. 

2. Even if we ran it at vocab load time not inside a compilation unit, it probably is still baked into the build bot image and saved. 

3. So then that’s why I made ``update-library``, but that would have to be called in a startup hook which would work except possibly if any library calls have to be made at vocab load time. 

4. The ``find-library`` machinery introduces new dependencies to deploy which if I remember on one of the platforms maybe macOS required reflection or one of the higher settings and bigger binary sizes. 

So, I think we should modify and maybe revert a little or test a lot. I made the mistake of doing a change like this for 0.97 and then releasing shortly after and broke the Linux users. This is better because it’s not fully qualified but still needs some good thinking if we plan on releasing shortly.

**Comment by bjourne** [comment]
:( Alright, I'll do a partial revert of that commit. But I think we should use it after the release. Referencing development libraries (like we do now) works for most Linux users, since they have root access to their machines and can install development packages. But if you are on a machine without root access (like if you are a student at a university) then you can't install development packages and won't be able to get factor running.

**Comment by mrjbq7** [comment]
I’m not trying to say we shouldn’t fix it, just that it’s a little more nuanced!

**Comment by mrjbq7** [comment]
Looks like we use find-library-from-list in opengl.gl on Linux. I think that usage isn’t quite correct for similar reasons to the previous list. 

Assuming deploy isn’t an issue, what are your thoughts on using update-library in a startup hook?

**Comment by bjourne** [comment]
At least on Linux it is not needed. :) All our linking problems will go away if we get Factor into the Linux distro's package repositories. It is something I'd like to work on after the 0.98 release. It will be way easier for Linux users to get factor running.

**Comment by mrjbq7** [comment]
We might need to solve the “conflicts with factor” binary/package name problem :-)
